versioned_hdf5/subchunk_map.py

```diff
@@ -45,113 +51,503 @@ def _smallest(x: Py_ssize_t, a: Py_ssize_t, m: Py_ssize_t) -> Py_ssize_t:
     return a + n * m
 
 
-@cython.cfunc
-def _subindex_chunk_slice(
-    s_start: Py_ssize_t,
-    s_stop: Py_ssize_t,
-    s_step: Py_ssize_t,
-    c_start: Py_ssize_t,
-    c_stop: Py_ssize_t,
-) -> slice:
-    """Given a slice(s_start, s_stop, s_step) indexing an axis of
-    a dataset, return the slice of the output array (on __getitem__)
-    or value parameter array (on __setitem__) along the
-    same axis that is targeted by the same slice after it's been
-    clipped to select only the data within the range [c_start, c_stop[
-
-    In other words:
-
-    a = _subindex_chunk_slice(s_start, s_stop, s_step, c_start, c_stop)
-    b = _subindex_slice_chunk(s_start, s_stop, s_step, c_start, c_stop)
-
-    __getitem__: out[a] = cache_chunk[b]
-    __setitem__: cache_chunk[b] = value[a]
-    """
-    start: Py_ssize_t = max(c_start, s_start)
-    # Get the smallest lcm multiple of common that is >= start
-    start = _smallest(start, s_start % s_step, s_step)
-    # Finally, we need to shift start so that it is relative to index
-    start = (start - s_start) // s_step
+class ChunkMapIndexError(IndexError):
+    """Raised by IndexChunkMapper.chunk_submap() if the requested range of chunk indices
+    does not intersect with IndexChunkMapper.chunk_indices.
 
-    stop: Py_ssize_t = min(c_stop, s_stop)
-    stop = _ceiling(stop - s_start, s_step) if stop > s_start else 0
+    We define a custom exception instead of IndexError as raising this is a normal
+    occourrence that will be caught quite a long way down the line and we don't want to
+    silence an accidental, genuine IndexError.
+    """
 
-    return slice(int(start), int(stop), 1)
 
+@cython.cclass
+class IndexChunkMapper:
```

Comment:
This class and all of its subclasses are just a refactor of the legacy code, with no logic changes.
I've rewired the legacy `as_subchunk_map` to use the refactored code.
There are minor additions, which are not used by the legacy `as_subchunk_map`:
- new methods `chunks_indexer` and `whole_chunks_indexer`
- the method `chunks_submap` allows selecting more than one contiguous chunk at a time, but `as_subchunk_map` will only ever ask for one
- the same method raises an exception for early exit when none of the chunks requested is impacted by the index. This never happens with the legacy `as_subchunk_map` by definition.